test/BinaryParse.cpp

```diff
@@ -24,12 +24,428 @@
+  os << "\nParsedInstruction( {";
+  spvtest::PrintTo(spvtest::WordVector(inst.words), &os);
+  os << "}, opcode: " << int(inst.opcode)
+     << " ext_inst_type: " << int(inst.ext_inst_type)
+     << " type_id: " << inst.type_id << " result_id: " << inst.result_id;
+  for (const auto& operand : inst.operands) {
+    os << " { offset: " << operand.offset << " num_words: " << operand.num_words
+       << " type: " << int(operand.type)
+       << " number_kind: " << int(operand.number_kind)
+       << " number_bit_width: " << int(operand.number_bit_width) << "}";
+  }
+  return os;
+}
+
+// Sanity check for the equality operator on ParsedInstruction.
+TEST(ParsedInstruction, ZeroInitializedAreEqual) {
+  spv_parsed_instruction_t pi = {};
+  ParsedInstruction a(pi);
+  ParsedInstruction b(pi);
+  EXPECT_THAT(a, ::testing::TypedEq<ParsedInstruction>(b));
+}
+
+// A template class with static member functions that forward to non-static
+// member functions corresponding to the header and instruction callbacks
+// for spvBinaryParse.
+template <typename Client, typename T>
+class BinaryParseClient : public T {
+ public:
+  static spv_result_t Header(void* user_data, spv_endianness_t endian,
+                             uint32_t magic, uint32_t version,
+                             uint32_t generator, uint32_t id_bound,
+                             uint32_t reserved) {
+    return static_cast<Client*>(user_data)
+        ->HandleHeader(endian, magic, version, generator, id_bound, reserved);
+  }
+
+  static spv_result_t Instruction(
+      void* user_data, const spv_parsed_instruction_t* parsed_instruction) {
+    return static_cast<Client*>(user_data)
+        ->HandleInstruction(parsed_instruction);
+  }
+};
+
+using Words = std::vector<uint32_t>;
+using Endians = std::vector<spv_endianness_t>;
+using Sentences = std::vector<Words>;  // Maybe this is too cute?
+using Instructions = std::vector<ParsedInstruction>;
+
+// A binary parse client that captures the results of parsing a binary,
+// and whose callbacks can be made to succeed for a specified number of
+// times, and then always fail with a given failure code.
+template <typename T>
+class CaptureParseResults
+    : public BinaryParseClient<CaptureParseResults<T>, T> {
+ public:
+  // Capture the header from the parser callback.
+  // If the number of callback successes has not yet been exhausted, then
+  // returns SPV_SUCCESS, which itself counts as a callback success.
+  // Otherwise returns the stored failure code.
+  virtual spv_result_t HandleHeader(spv_endianness_t endian, uint32_t magic,
+                                    uint32_t version, uint32_t generator,
+                                    uint32_t id_bound, uint32_t reserved) {
+    endians_.push_back(endian);
+    headers_.push_back({magic, version, generator, id_bound, reserved});
+    return ComputeResultCode();
+  }
+
+  // Capture the parsed instruction data from the parser callback.
```

Comment:
Optional: consider using googlemock for this.  Simply declare `HandleInstruction` a `MOCK_METHOD1` and use `EXPECT_CALL` in the tests below.  The framework is powerful enough that you shouldn't need any of the rest of the class.